### ISSUE TITLE
Some questions about the data and the training process

### ISSUE THREAD

**Issue by ahikaml** [anicolson/DeepXi#44]
Hi,
First of all let me express my appreciation for this project. Very interesting.
I have some questions regarding the data and the training process:

1. I see that most of the clean speech utterances in _train-clean-100_ and in _VCTK_ are shorter than the receptive field size of the network. Does it mean that this specific databases don't exploit the full potential of the network?

2. Under the assumption that the network inference is causal (up to one buffer) - does it mean that the SNR estimation in the first seconds is expected to be less accurate due to the short respective field?  I mean to the fact that we don't have enough information about the past, similar to a linear filter that has some "delay time" even if it's a minimum phase.   

3. I noticed that when I use short clean speech files (up to 8 sec) for training this "delay time" is shorter than when I use longer files (~up to 20sec). Is it expected? The files in both cases are from the same databases but in the first case I just split them to shorter parts.

4. Regarding the files normalization  - I noticed that you normalize the noise files according to  |s|/(|d|*SNR). How sparse noise files should be normalized? The problem is that |d| is very depends on the sparsity of the signal. For example the RMS  of noise file containing a slamming door noise every 3 seconds is much lower than the RMS of some constant noise (fan, car, airplane) because in the first case most of the file is a silence. I'm asking because all of the sparse files were created in higher SNR due to very low value of |d|.
     
4. I saw that the databases in the link contain only a single speaker in each file. If I'd like to test cases of several speakers (not simultaneously ) in the same file, similar to a real life conversation, does it mean that I need to train the network with several speaker in each file as well?
In other words, does the network study specific speaker characteristics or more general characteristics relevant to the human voice?

Thank you
Ahikam

**Comment by anicolson** [comment]
Hi,

For 1., yes, a training set with longer files should be used. What would be interesting is to see a histogram of the durations of all the clean speech files in the training set.

For 2, certainly, the network has less information to infer from. 

For 3, for the split files, is the speech closer to the start of the file than the files that are not split?

For 4, the noise files are scaled in the time domain, not frequency domain. This is certainly an issue that can be looked into to improve the framework.

For 5, the network has been trained with only a single speaker as the target, so it may struggle with multiple speakers.

Hope this helps,
Aaron.

**Comment by ahikaml** [comment]
Hi,

Thanks for your response. It's very helpful.

Regarding (1) see for example the VCTK histogram below. Most of the files are shorter than the net receptive field.
In the other hand, as I said, somehow the conversion time in the inference is lower when the net is trained on shorter files (up to 2 secs). 

Regarding (3) - No. In long files I just cut any extra silence in the beginning to make the learning as effective as possible. 

Regarding (4) - I assumed that scaling in the time domain is identical to scaling in the frequency domain (Parseval's theorem). If you meant to a different scaling factor per frequency bin then it's not different than equalization.  
Let me know if you think I'm wrong.   
Anyway I thought that scaling by the maximum or by some upper percentile might give a decent solution for this issue but I wasn't really sure. I guess I'll have to try...

One more question - is there a good way to know, after the training, what is the effective receptive field size? I mean by looking on the coefficient values of any specific layer? Could it be related to the coefficients values in the residual blocks with the highest dilation ratio?

Thanks
Ahikam

![DurationHistVCTK](https://user-images.githubusercontent.com/32457828/86958926-9db58c80-c165-11ea-921a-77e9e12e67f9.JPG)

**Comment by anicolson** [comment]
So with 1, just for your information, the train-clean-100 set from librispeech has a minimum duration of 1.4 seconds, a mean duration of 12.3 seconds and a maximum duration of 17.2. 

This can be helpful to calculate the receptive field: https://fomoro.com/research/article/receptive-field-calculator#

Hope this helps, 
Aaron

**Comment by ahikaml** [comment]
Thanks for the link. It seems that this site helps you to calculate the theoretical receptive field, i.e. the maximum that the network can "see" back to the past . 
However to me it feels that sometimes, depends on the training data, the network can "choose" not to look on such a long history even though its architecture supports that. 
For example when the network has receptive field of 6 sec but it learn that information of more than 2 seconds back to the past isn't relevant to the current SNR.
 
In such case I believe that its coefficients will be updated accordingly to represent this short receptive field. 
So.. the question is: can I recognized such situations by looking on the network coefficient after the training, or it's totally a black box and all we can do is just checking if the 'a priory' SNR we got from the net is accurate enough?

Thanks
Ahikam
